Sources/armory/renderpath/RenderPathDeferred.hx

```diff
@@ -730,6 +751,97 @@ class RenderPathDeferred {
 		}
 		#end
 
+		#if rp_ssrefr
+		{
+			if (armory.data.Config.raw.rp_ssrefr != false)
+			{
+				path.setTarget("gbufferD1");
+				path.bindTarget("_main", "tex");
+				path.drawShader("shader_datas/copy_pass/copy_pass");
+
+				path.setTarget("refr");
+				path.bindTarget("tex", "tex");
+				path.drawShader("shader_datas/copy_pass/copy_pass");
+
+				RenderPathCreator.setTargetMeshes();
+				path.drawMeshes("refraction");
+
+				path.setTarget("tex");
+				path.bindTarget("_main", "gbufferD");
+				path.bindTarget("gbuffer0", "gbuffer0");
+				path.bindTarget("gbuffer1", "gbuffer1");
+
+				#if rp_gbuffer2
+				{
+					path.bindTarget("gbuffer2", "gbuffer2");
+				}
+				#end
+
+				#if rp_gbuffer_emission
+				{
+					path.bindTarget("gbuffer_emission", "gbufferEmission");
+				}
+				#end
+
+				#if (rp_ssgi != "Off")
+				{
+					if (armory.data.Config.raw.rp_ssgi != false) {
+						path.bindTarget("singlea", "ssaotex");
+					}
+					else {
+						path.bindTarget("empty_white", "ssaotex");
+					}
+				}
+				#end
+
+				var voxelao_pass = false;
+				#if rp_voxelao
+				if (armory.data.Config.raw.rp_voxels != false)
+				{
+					#if arm_config
+					voxelao_pass = true;
+					#end
+					path.bindTarget(voxels, "voxels");
+					#if arm_voxelgi_temporal
+					{
+						path.bindTarget(voxelsLast, "voxelsLast");
+					}
+					#end
+				}
+				#end
+
+				#if rp_shadowmap
+				{
+					#if arm_shadowmap_atlas
+					Inc.bindShadowMapAtlas();
+					#else
+					Inc.bindShadowMap();
+					#end
+				}
+				#end
+
+				#if rp_material_solid
+				path.drawShader("shader_datas/deferred_light_solid/deferred_light");
+				#elseif rp_material_mobile
+				path.drawShader("shader_datas/deferred_light_mobile/deferred_light");
+				#else
+				voxelao_pass ?
+					path.drawShader("shader_datas/deferred_light/deferred_light_VoxelAOvar") :
+					path.drawShader("shader_datas/deferred_light/deferred_light");
+				#end
```

Comment:
Is it really required to re-run the entire deferred lighting pass? This looks slow

Reply:
Hi, thanks for reviewing.
When doing refraction, we need to render the scene without the refractive objects, having done so, we render the refractive meshes using the data from the first pass.
The reason why we are doing another deferred pass is because we want the refractive objects to be shaded.

I think it's ok to do so since the buffers are filled only with the data from the refractive objects on the second pass.

I'm going to fix the various errors this morning.

Reply:
> Is it really required to re-run the entire deferred lighting pass? This looks slow

Yes, objects must get shaded. Nonetheless, buffers only countaint data for the refractive objects. I don't know what the deferred shader does in this case but if there is a performance impact, we should add a mask and discard the fragment.